### ISSUE TITLE
Multiple Scenarios and Browserstack

### ISSUE THREAD

**Issue by ttretau** [jan-molak/serenity-js#34]
Hi there,
I am using Serenity-JS fine with local Webdriver and with Selenium Grid - with multiple Features and Scenarios.
Starting to test with Browserstack cloud service only the first Scenario is finished. 
After that all scenarios return with
"_Step Definition: node_modules/serenity-js/src/serenity-cucumber/stage_cue.ts:6
   Message:
     function timed out after 10000 milliseconds_"
or
"_Step Definition: node_modules/serenity-js/src/serenity-cucumber/stage_cue.ts:6
   Message:
     NoSuchSessionError: This driver instance does not have a valid session ID (did you call WebDriver.quit()?) and may no longer be used._"

Using Version 1.2.1 and Protractor 5.1.1. Could this be related to https://github.com/angular/protractor/issues/4087?
Have you tested this with browserstack? 

Kind regards,
Torben

**Comment by ttretau** [comment]
Ah, think I fixed this by changing protractor.conf, like here:
https://github.com/serenity-js/tutorial-from-scripts-to-serenity/blob/master/protractor.conf.js#L9
I had 'serenity-js' before there..

**Comment by jan-molak** [comment]
Hey @ttretau, Serenity/JS itself is tested with BrowserStack, so yes, those two should work together.

Could you please provide a minimal test case demonstrating the issue you're seeing?

Thanks,
Jan

**Comment by ttretau** [comment]
Hey @jan-molak,
Ok, I updated to latest dependencies (like the tutorial-project). Now I got
_[09:18:38] E/launcher - BUG: launcher exited with 1 tasks remaining_
I will try to build a minimal example project, reproducing that..

**Comment by ttretau** [comment]
I created this Example:
[more or less running, with not so well session behaviour](https://github.com/ttretau/serenity-js-example/tree/browserstack_running_more_or_less)
After [updating dependencies](https://github.com/ttretau/serenity-js-example/tree/browserstack_stopping) it stopped after first scenario with
[11:58:44] E/launcher - BUG: launcher exited with 1 tasks remaining

**Comment by jan-molak** [comment]
Thanks @ttretau, I managed to reproduce the issue.

The `[11:58:44] E/launcher - BUG: launcher exited with 1 tasks remaining` is caused by the `protractor-cucumber-framework`, which [both](https://github.com/ttretau/serenity-js-example/blob/68407e1cf9626672f0914847af31d07cb6bb6b14/protractor.conf.js#L61) the  [examples](https://github.com/ttretau/serenity-js-example/blob/browserstack_stopping/protractor.conf.js#L61) are using instead of `serenity-js`.

Having said that, when the example is changed to use `serenity-js`, the actual issue is triggered: `function timed out after 10000 milliseconds`. 
It seems like 10 seconds is not long enough to allow Serenity/JS to pull the screenshots from BrowserStack, so the timeout allowed for the cleanup tasks needs to be extended and ideally made configurable so that you can tweak it yourself.

Thanks for reporting this problem, a fix is on the way.

Jan

**Comment by ttretau** [comment]
Hey @jan-molak, great news! Thanks for the prompt answer and the hint with wrong configuration. Let me know if I can help you fixing/testing the bug.. 
Torben

**Comment by jan-molak** [comment]
Hey @ttretau, upgrading to Serenity/JS 1.2.4 [makes the example pass](https://github.com/jan-molak/serenity-js-example/commit/e2777d2bdc416c4b68f228957392e9dd520e8391). I'm looking into a nicer way of managing timeouts now.

**Comment by ttretau** [comment]
Cool, now it works! Great work!

**Comment by ttretau** [comment]
Hi @jan-molak, 
I tried your new config with the newest serenity-js version (1.3.0).
In practice and with [this simple example](https://github.com/ttretau/serenity-js-example) it shows some problems handling multiple scenarios and their according sessions in browserstack.
The first scenario is handled fine with the corresponding successful browserstack session. After that it starts another short lived session, doing "nothing" except setting a timeout.
The next scenario does not finish and is running in a timeout. After that another "empty" browserstack session is started. 

Looks like it also happening running locally with this example. leaving open windows until timeout..
Let me know if/how I can support you in finding/fixing this issue. 
Kind regards,
Torben

**Comment by jan-molak** [comment]
Hey @ttretau!

I cloned the example repository and run it against browserstack with the following result:

```
[22:59:10] I/hosted - Using the selenium server at http://hub-cloud.browserstack.com/wd/hub
Feature: Add new items to the todo list

    In order to avoid having to remember things that need doing
    As a forgetful person
    I want to be able to record what I need to do in a place where I won't forget about them

  Scenario: Adding an item to a list with other items
  ✔ Given that James has a todo list containing Buy some cookies, Walk the dog
  ✔ When he adds Buy some cereal to his list
  ✔ And he adds Buy some cereal to his list
  ✔ And he adds Buy some cereal to his list
  ✔ Then his todo list should contain Buy some cookies, Walk the dog, Buy some cereal, Buy some cereal, Buy some cereal

  Scenario: Adding an item to a list with other items with Umlauts
  ✔ Given that Klaus Jäger has a todo list containing Kaufe zwei Mehlsäcke, geh das Auto waschen
  ✔ When he adds Kaufe vier Bananen to his list
  ✔ And he adds Kaufe fünf Gurken to his list
  ✔ And he adds Kaufe fünf Gurken to his list
  ✔ Then his todo list should contain Kaufe zwei Mehlsäcke, geh das Auto waschen, Kaufe vier Bananen, Kaufe fünf Gurken, Kaufe fünf Gurken

2 scenarios (2 passed)
10 steps (10 passed)
0m37.711s
[23:00:27] I/launcher - 0 instance(s) of WebDriver still running
[23:00:27] I/launcher - Chrome #01 passed
```

What exact parameters are you running the tests with, please?

J

**Comment by ttretau** [comment]
Hi,
I am running the tests exactly this way. My result 

```
[08:26:08] I/launcher - Running 1 instances of WebDriver
[08:26:08] I/hosted - Using the selenium server at http://hub-cloud.browserstack.com/wd/hub
Feature: Add new items to the todo list
[...] 
2 scenarios (2 passed)
10 steps (10 passed)
0m54.357s
[..here it stays for 4 minutes..]
[08:30:59] I/launcher - 0 instance(s) of WebDriver still running
[08:30:59] I/launcher - Chrome #01 passed
```

Same as yours, except 4 minutes pause until the WebDriver output..
Leading to this Browserstack Sessions:
```
Duration: 4 mins 40 secs
Session	OS	Browser / Device	Duration	Finished
 single_test	 Win 10	 Chrome 57.0	21s	a min ago
 single_test	 Win 10	 Chrome 57.0	1m 41s	1m ago
 single_test	 Win 10	 Chrome 57.0	1m 56s	2m ago
 single_test	 Win 10	 Chrome 57.0	10s	4m ago
 single_test	 Win 10	 Chrome 57.0	29s	4m ago

```
As it works fine on your side I think I have to look for a Browserstack or Network problem..

**Comment by ttretau** [comment]
Ah, perhaps queuing on Browserstack - we have only one parallel tests available - so although maxInstances and count are set to one on protractor side something is put into a queue on Browserstack-side.. Do you have more than one parallel tests available in your Browserstack-Account?

**Comment by jan-molak** [comment]
Hey @ttretau, it does sound like queueing on BrowserStack; I have 2 parallel tests available, so running with two browsers works fine. I suppose it's sensible to set `maxInstances` to a number of browsers the server can actually spawn.

Does that answer your question?

**Comment by ttretau** [comment]
Hi @jan-molak, i debugged a little into browser starts and it looks wired to me - this is my debug output:
```
BLUB enter the stage [object Object]
Feature: Add new items to the todo list

    In order to avoid having to remember things that need doing
    As a forgetful person
    I want to be able to record what I need to do in a place where I won't forget about them

  Scenario: Adding an item to a list with other items
BLUB browse the web http://todomvc.com
BLUB browse the web http://todomvc.com
  ✔ Given that James has a todo list containing Buy some cookies, Walk the dog
  ✔ When he adds Buy some cereal to his list
  ✔ And he adds Buy some cereal to his list
  ✔ And he adds Buy some cereal to his list
  ✔ Then his todo list should contain Buy some cookies, Walk the dog, Buy some cereal, Buy some cereal, Buy some cereal

  Scenario: Adding an item to a list with other items with Umlauts
BLUB browse the web http://todomvc.com
BLUB browse the web http://todomvc.com
  ✔ Given that Klaus Jäger has a todo list containing Kaufe zwei Mehlsäcke, geh das Auto waschen
```
The "browse the web debug" output is in [BrowseTheWeb.using(), instatiating the BrowseTheWeb Ability](https://github.com/jan-molak/serenity-js/blob/master/packages/serenity-js/src/serenity-protractor/screenplay/abilities/browse_the_web.ts#L19). Does it mean that it is starting too much browser-windows for a scenario?

**Comment by jan-molak** [comment]
Interesting; 

The output seems weird; `BrowseTheWeb` is not itself responsible for starting the browser, it receives an instantiated `browser` object from Protractor, which is also responsible for sharding the tests and running them in parallel...

Can you push your debug version to github so that I can try to reproduce?

**Comment by ttretau** [comment]
Ah, I now saw, that if I run this sample locally it flickers after the actual tested browser window opens, so I think there is also more than one window opened if you run it without Browserstack. 
My debug version contains only the `console.log` in `BrowseTheWeb.using()` to show that how often it is called..
I am looking into how protractor browser is instantiated..

**Comment by jan-molak** [comment]
Cool, please keep me posted!

**Comment by ttretau** [comment]
Hey @jan-molak , 
today I found out that this could depend on the "default" crew configuration in serenity. If I comment out [these lines](https://github.com/jan-molak/serenity-js/blob/master/packages/serenity-js/src/serenity-protractor/framework/serenity_protractor_framework.ts#L40) it works. If they are in, after the actual session another 3 browser sessions are started.. I am trying to get into how that should work..

**Comment by jan-molak** [comment]
Hmm, the `ProtractorReporter` and `ProtractorNotifier` take the instance of Protractor's `Runner`. Is it one of those objects that cause trouble? I don't think the `StandInd` should affect anything ...

**Comment by ttretau** [comment]
That's right. `StandIns` only leads to 'stand-in' actor entering the stage. 
So the `Runner` instances look hot. The Notifier/Reporter are spawning new sessions?

**Comment by jan-molak** [comment]
No, they shouldn't. The idea here is that Protractor does all the heavy lifting and session management and Serenity/JS provides an abstraction layer on top of that.

[Notifier](https://github.com/jan-molak/serenity-js/blob/master/packages/serenity-js/src/serenity-protractor/reporting/protractor_notifier.ts) invokes `afterEach` to let Protractor know when the test is finished (as per [the requirements](https://github.com/angular/protractor/blob/master/lib/frameworks/README.md#future-requirements)).

[Reporter](https://github.com/jan-molak/serenity-js/blob/master/packages/serenity-js/src/serenity-protractor/reporting/protractor_reporter.ts) notifies protractor of [test passes/failures](https://github.com/jan-molak/serenity-js/blob/master/packages/serenity-js/src/serenity-protractor/reporting/protractor_reporter.ts#L67) and the final result ([protractor spec](https://github.com/angular/protractor/blob/master/lib/frameworks/README.md#requirements)).

**Comment by ttretau** [comment]
Ah, thanks. That explanation brings some light.. So I think getting deeper into protractor debugging makes sense..

**Comment by ttretau** [comment]
Nothing new.. :-( Did not found out anything regarding this problem.. Looks like protractor integration is working fine.. After scenario is finished a short browser session is popping up. Could that also be a problem with Cucumber?

**Comment by jan-molak** [comment]
Hmm, and what if you don't instruct Protractor to restart the browser between the tests?
Maybe the short session is caused by the restart? 🤔

**Comment by ttretau** [comment]
Ah, you nailed it! :+1: That is a workaround for to the browserstack problem..
Could this be related to this protractor problem? https://github.com/angular/protractor/issues/2001

**Comment by jan-molak** [comment]
@ttretau it looks like the problem is not related to BrowserStack, but more to how Protractor handles browser restarts. I've created a new ticket so we can have a more focused conversation there. #56
